spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/ModuleDefinition.java

```diff
@@ -0,0 +1,417 @@
+			this.parameters.put(name, value);
+			return this;
+		}
+
+		/**
+		 * Add the contents of the provided map to the map of module parameters.
+		 *
+		 * @param parameters module parameters
+		 * @return this builder object
+		 *
+		 * @see ModuleDefinition#parameters
+		 */
+		public Builder addParameters(Map<String, String> parameters) {
+			this.parameters.putAll(parameters);
+			return this;
+		}
+
+		/**
+		 * Set a module binding.
+		 *
+		 * @param name binding name
+		 * @param value binding value
+		 * @return this builder object
+		 *
+		 * @see ModuleDefinition#bindings
+		 */
+		public Builder setBinding(String name, String value) {
+			this.bindings.put(name, value);
+			return this;
+		}
+
+		/**
+		 * Add the contents of the provided map to the map of module bindings.
+		 *
+		 * @param bindings module bindings
+		 * @return this builder object
+		 *
+		 * @see ModuleDefinition#bindings
+		 */
+		public Builder addBindings(Map<String, String> bindings) {
+			this.bindings.putAll(bindings);
+			return this;
+		}
+
+		/**
+		 * Return name of module. Typically this module is present under
+		 * {@code $XD_HOME/modules/[module type]}.
+		 *
+		 * @return module name
+		 */
+		public String getModuleName() {
+			return moduleName;
+		}
+
+		/**
+		 * Return symbolic name of a module. This may be explicitly specified in
+		 * the DSL string, which is required if the stream contains another module
+		 * with the same name. Otherwise, it will be the same as the module name.
+		 *
+		 * @return module label
+		 */
+		public String getModuleLabel() {
+			return moduleLabel;
+		}
+
+		/**
+		 * Return name of deployable unit this module instance belongs to
+		 * (such as a stream or job).
+		 *
+		 * @return group name
+		 */
+		public String getGroup() {
+			return group;
+		}
+
+		/**
+		 * Return position in stream/job definition relative to the other modules in
+		 * the definition. 0 indicates the first/leftmost position.
+		 *
+		 * @return module index
+		 */
+		public int getIndex() {
+			return index;
+		}
+
+		/**
+		 * Return parameters for module. This is specific to the type of module -
+		 * for instance an http module would include a port number as a parameter.
+		 * <br />
+		 * Note that the contents of this map are <b>mutable</b>.
+		 *
+		 * @return map of module parameters
+		 */
+		public Map<String, String> getParameters() {
+			return parameters;
+		}
+
+		/**
+		 * Return bindings for module. Note that the contents of this map are <b>mutable</b>.
+		 *
+		 * @return map of module bindings
+		 */
+		public Map<String, String> getBindings() {
+			return bindings;
+		}
+
+		/**
+		 * Create a {@code Builder} object pre-populated with the configuration
+		 * for the provided {@link ModuleDefinition}.
+		 *
+		 * @param definition module descriptor
+		 * @return pre-populated builder object
+		 */
+		public static Builder fromModuleDefinition(ModuleDefinition definition) {
```

Comment:
do we still need this?... seems that we now should only support a fully populated ModuleDefinition, whereas this method I believe was originally used for the case where depending on its context and prior usage a definition may not have contained all of the properties as needed in another context (e.g. declaration vs. deployment).